contracts/TrancheWrapper.sol

```diff
@@ -0,0 +1,256 @@
+    function redeem(
+        uint256 shares,
+        address receiver,
+        address owner
+    ) external nonReentrant returns (uint256) {
+        if (shares == type(uint256).max) shares = balanceOf(owner);
+
+        (uint256 _withdrawn, uint256 _burntShares) = _redeem(shares, receiver, owner);
+
+        emit Withdraw(msg.sender, receiver, owner, _withdrawn, _burntShares);
+        return _withdrawn;
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                    DEPOSIT/WITHDRAWAL LIMIT LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @dev Returns the maximum amount of the underlying asset that can be deposited into the Vault for the receiver,
+     * through a deposit call.
+     */
+    function maxDeposit(
+        address /* receiver */
+    ) public view returns (uint256) {
+        IdleCDO _idleCDO = idleCDO;
+
+        uint256 _depositLimit = _idleCDO.limit(); // TVL limit in underlying value
+        uint256 _totalAssets = _idleCDO.getContractValue(); // TVL in underlying value
+        if (_depositLimit == 0) return type(uint256).max; // 0 means unlimited
+        if (_totalAssets >= _depositLimit) return 0;
+        return _depositLimit - _totalAssets;
+    }
+
+    function maxMint(address receiver) external view returns (uint256) {
+        uint256 _maxDeposit = maxDeposit(receiver);
+        if (_maxDeposit == type(uint256).max) return type(uint256).max;
+        return convertToShares(_maxDeposit);
+    }
+
+    function maxWithdraw(address owner) external view returns (uint256) {
+        bool withdrawable = isAATranche ? idleCDO.allowAAWithdraw() : idleCDO.allowBBWithdraw();
+        if (!withdrawable) return 0;
+        return convertToAssets(balanceOf(owner));
+    }
+
+    function maxRedeem(address owner) external view returns (uint256) {
+        bool withdrawable = isAATranche ? idleCDO.allowAAWithdraw() : idleCDO.allowBBWithdraw();
+        if (!withdrawable) return 0;
+        return balanceOf(owner);
+    }
+
+    /// @notice Deposit underlying tokens into IdleCDO
+    /// @dev This function SHOULD be guarded to prevent potential reentrancy
+    /// @param amount Amount of underlying tokens to deposit
+    /// @param receiver receiver of tranche shares
+    /// @param depositor depositor of underlying tokens
+    function _deposit(
+        uint256 amount,
+        address receiver,
+        address depositor
+    ) internal returns (uint256 deposited, uint256 mintedShares) {
+        IdleCDO _idleCDO = idleCDO;
+        IERC20 _token = IERC20(token);
+
+        SafeERC20.safeTransferFrom(_token, depositor, address(this), amount);
+
+        if (_token.allowance(address(this), address(_idleCDO)) < amount) {
+            _token.approve(address(_idleCDO), 0); // Avoid issues with some tokens requiring 0
+            _token.approve(address(_idleCDO), type(uint256).max); // Vaults are trusted
+        }
```

Comment:
Is this really needed? can we simply set the allowance on the constructor/initialize method given that idleCDO contract is trusted?